aten/src/ATen/native/cuda/NativeFunctionsCuda.cu

```diff
@@ -0,0 +1,210 @@
+#include "ATen/native/NativeFunctions.h"
+#include <cfloat>
+
+namespace at {
+namespace native {
+
+__host__ __device__ __forceinline__ float fmin(float a, float b) {
+  return a > b ? b : a;
+}
+
+__host__ __device__ __forceinline__ float fmax(float a, float b) {
+  return a > b ? a : b;
+}
+
+template <typename T>
+__global__ void RoiPooling2d_forward_kernel(
+  const int outputElements,
+  const T *input,
+  const T *rois,
+  const T spatialScale,
+  const int inputChannels,
+  const int inputHeight,
+  const int inputWidth,
+  const int pooledHeight,
+  const int pooledWidth,
+  T *output,
+  int *argmaxes)
+{
+  for (int linearIndex = blockIdx.x * blockDim.x + threadIdx.x;
+       linearIndex < outputElements;
+       linearIndex += blockDim.x * gridDim.x)
+  {
+    // Calculate position in output Tensor, i.e. a specific combination
+    // of proposal, channel, pool height and pool width
+    // TODO: write to improve performance by minimize computation
+    int pw = linearIndex % pooledWidth;
+    int ph = (linearIndex / pooledWidth) % pooledHeight;
+    int ch = (linearIndex / pooledWidth / pooledHeight) % inputChannels;
+    int proposal = linearIndex / pooledWidth / pooledHeight / inputChannels;
+
+    // Get particular proposal data
+    const T *roisOffset = rois + (proposal * 5);
+    int n = roisOffset[0];
+    int startWidth = llrintf(roisOffset[1] * spatialScale);
+    int startHeight = llrintf(roisOffset[2] * spatialScale);
+    int endWidth = llrintf(roisOffset[3] * spatialScale);
+    int endHeight = llrintf(roisOffset[4] * spatialScale);
+
+    // TODO: fix malformed RoIs to be 1x1
+
+    int roiHeight = endHeight - startHeight;
+    int roiWidth = endWidth - startWidth;
+
+    // Calculate size of tile based on the size of this particular RoI and the
+    // output size
+    T tileHeight = static_cast<T>(roiHeight) / static_cast<T>(pooledHeight);
+    T tileWidth = static_cast<T>(roiWidth) / static_cast<T>(pooledWidth);
+
+    // Calculate offset into the pooled region
+    int tileHStart = static_cast<int>(floorf(static_cast<T>(ph) * tileHeight));
+    int tileWStart = static_cast<int>(floorf(static_cast<T>(pw) * tileWidth));
+    int tileHEnd = static_cast<int>(ceilf(static_cast<T>(ph + 1) * tileHeight));
+    int tileWEnd = static_cast<int>(ceilf(static_cast<T>(pw + 1) * tileWidth));
```

Comment:
BTW using floats for indexing is a risky thing. I think it's better to coordinate integer multiplication/division to get floor and ceil out of the box. Here, you could replace it with this:
```cpp
int tileHStart = (ph * roiHeight) / pooledHeight;
...
// NOTE: ((ph + 1) * roiHeight + roiHeight - 1) == ((ph + 2) * roiHeight - 1)
int tileHEnd = ((ph + 2) * roiHeight - 1) / pooledHeight; 
```

Reply:
(to highlight one of the risks: 32-bit addressable tensors aren't addressable with floats that only have 23 bits of mantissa)